packages/zksend/src/experimental/wallet/channel/index.ts

```diff
@@ -0,0 +1,181 @@
+
+export { StashedRequest, StashedResponse };
+
+export class StashedPopup {
+	#popup: Window;
+
+	#version: string;
+	#id: string;
+	#origin: string;
+	#name: string;
+	#chain: string | undefined;
+
+	#promise: Promise<unknown>;
+	#resolve: (data: unknown) => void;
+	#reject: (error: Error) => void;
+
+	#interval: ReturnType<typeof setInterval> | null = null;
+
+	constructor({
+		name,
+		origin = DEFAULT_STASHED_ORIGIN,
+		chain,
+	}: {
+		name: string;
+		origin?: string;
+		chain?: string;
+	}) {
+		const popup = window.open('about:blank', '_blank');
+
+		if (!popup) {
+			throw new Error('Failed to open new window');
+		}
+		this.#popup = popup;
+
+		this.#id = crypto.randomUUID();
+		this.#origin = origin;
+		this.#name = name;
+		this.#version = 'v1';
+		this.#chain = chain;
+		const { promise, resolve, reject } = withResolvers();
+		this.#promise = promise;
+		this.#resolve = resolve;
+		this.#reject = reject;
+
+		this.#interval = setInterval(() => {
+			try {
+				if (this.#popup.closed) {
+					this.#cleanup();
+					reject(new Error('User closed the Stashed window'));
+				}
+			} catch {
+				// This can error during the login flow, but that's fine.
+			}
+		}, 1000);
+	}
+
+	send<T extends StashedRequestData['type']>({
+		type,
+		...data
+	}: {
+		type: T;
+	} & Extract<StashedRequestData, { type: T }>): Promise<StashedResponseTypes[T]> {
+		window.addEventListener('message', this.#listener);
+
+		const requestData = {
+			...data,
+			version: this.#version,
+			requestId: this.#id,
+			appOrigin: window.origin,
+			appName: this.#name,
+			chain: this.#chain,
+		};
+
+		const encodedRequestData = btoa(JSON.stringify(requestData));
+
+		this.#popup.location.assign(
+			`${this.#origin}/dapp/${type}${data ? `#${encodedRequestData}` : ''}`,
+		);
+
+		return this.#promise as Promise<StashedResponseTypes[T]>;
+	}
+
+	close() {
+		this.#cleanup();
+		this.#popup.close();
+	}
+
+	#listener = (event: MessageEvent) => {
+		if (event.origin !== this.#origin) {
+			return;
+		}
+		const { success, output } = safeParse(StashedResponse, event.data);
+		if (!success || output.id !== this.#id) return;
+
+		this.#cleanup();
+
+		if (output.payload.type === 'reject') {
+			this.#reject(new Error('User rejected the request'));
+		} else if (output.payload.type === 'resolve') {
+			this.#resolve(output.payload.data);
+		}
+	};
+
+	#cleanup() {
+		if (this.#interval) {
+			clearInterval(this.#interval);
+			this.#interval = null;
+		}
+		window.removeEventListener('message', this.#listener);
+	}
+}
+
+export class StashedHost {
+	#request: InferOutput<typeof StashedRequest>;
+
+	constructor(request: InferOutput<typeof StashedRequest>) {
+		if (typeof window === 'undefined' || !window.opener) {
+			throw new Error(
+				'StashedHost can only be used in a window opened through `window.open`. `window.opener` is not available.',
+			);
+		}
+
+		this.#request = request;
+	}
+
+	static fromUrl(url: string = window.location.href) {
```

Comment:
We probably should make this fromPayload so that the host can just pass the big blob instead of needing to exactly match the URL. Gives us more flexibility